File: 125-is-palindrome/solution.go
package main

import (
	"fmt"
	"strings"
)

func isPalindrome(s string) bool {
	s = strings.ToLower(s)
	l, r := 0, len(s)-1
	for l < r {
		for l < r && !isalnum(s[l]) {
			l++
		}
		for l < r && !isalnum(s[r]) {
			r--
		}
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}
	return true
}
func isalnum(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9')
}
func main() {
	s := "A man, a plan, a canal: Panama"
	fmt.Println("Ans = ", isPalindrome(s))
}

File: 125-is-palindrome/solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not s[l].isalnum():
                l += 1
            while l < r and not s[r].isalnum():
                r -= 1
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True
if __name__ == "__main__":
    s = "A man, a plan, a canal: Panama"
    print("Ans = ", Solution().isPalindrome(s))

#如果字串是字母數字字串，則傳回 True，否則傳回 False。